### ISSUE TITLE
[RFC] Add support for distributed full-graph training and inference

### ISSUE THREAD

**Issue by hesham-mostafa** [dmlc/dgl#5084]
## 🚀 Feature
This issue is for discussing the integration of the Sequential Aggregation and Rematerialization technique for distributed full-graph training and inference from [here](https://github.com/IntelLabs/SAR) into DGL.

## Motivation
For large graphs, DGL supports distributed sampling-based training (DistDGL) where the graph is partitioned across several machines and distributed sampling takes care of creating the mini-batches for the trainers. There is no mechanism, however, to do either distributed full-graph training or distributed full-graph inference. Full-graph inference would be useful to get test/validation accuracies after doing sampling-based training. While you could do GNN inference using graph samples, the results would have some randomness. Taking the average predictions of several inference runs is possible but costly. Moreover, there is no guarantee that you are not discarding some crucial structural or node features.

There are many graph sampling methods out there, and some are problem specific (in fraud detection models for example), which would require new users to experiment to find the most performant method. Choosing the best sampling hyperparameters (such as neighbor sampling fanout) is also not trivial. Conceptually and in terms of hyper-parameter counts, full-graph training is much simpler. It would be beneficial to be able to train on the full graph (no sampling involved) as a sanity check and also to obtain a useful baseline to compare against the performance of various sampling-based training methods. 

The proposed implementation uses the graph partitions creates by DGL's metis (``dgl.distributed.partition_graph``) to construct a distributed graph object at each worker. This distributed graph object can be used as a drop-in replacement to ``dgl.DGLHeterograph`` and passed to a GNN. Under the hood, we take care of communicating the features of nodes at the partition boundaries in the forward pass, and collecting their gradients in the backward pass to realize a full-graph forward and backward GNN pass. We guarantee that no worker would need to materialize more than 2 graph partitions at once (even for densely connected graphs), and therefore obtain linear memory scaling, i.e, per-worker node/edge feature memory goes down linearly with number of workers.

## Alternatives

There is currently no simple way to implement distributed full-graph training/inference in DGL. 

## Pitch
Add the distributed full-graph training scheme from this repo [here](https://github.com/IntelLabs/SAR) to DGL. The repo implements the sequential aggregation and rematerialization scheme from the paper [here](https://proceedings.mlsys.org/paper/2022/hash/5fd0b37cd7dbbb00f97ba6ce92bf5add-Abstract.html) and wraps the entire functionality into a DGLHeterograph-like class that runs distributed full-graph forward and backward passes under the hood.

## Additional context
For distributed full-graph inference, we can distribute the graph across several machines. However, when doing the message passing in a GNN layer, a machine might need to fetch a substantial part of the graph to obtain the neighborhood of its local nodes as shown below. This can exhaust the machine's memory. This problem is more severe in more densely connected graphs. 

![image](https://user-images.githubusercontent.com/16276103/209868974-94dddf28-234d-4109-858c-29918db2a596.png)

The simple solution depends on the fact that the majority of GNN reduction methods such as ``sum``, ``mean`` and ``max`` can be recursively applied, i.e, ``reduce(A,B,C) == reduce(A,reduce(B,C))``, which would allow us to do the message passing sequentially. Even functions as ``edge_softmax`` used in attention-based GNNs can be calculated sequentially. The sequential  message passing operation is shown below:

![image](https://user-images.githubusercontent.com/16276103/209868994-0163d9fc-75a3-4dce-917d-461ce5352c1e.png)

After fetching the neighboring nodes from only one remote worker and aggregating their messages, machine 1 can delete these remote nodes and go on to fetch the remote node features from the next worker. We thus obtain linear memory scaling in the full-graph inference pass, as each machine would at most need to materialize two graph partitions, its own partition and the remote partition whose messages are being actively aggregated. 

The situation becomes more complicated when doing full-graph training, as even if we fetch and aggregate messages from remote partitions one by one, the remotely fetched node features would live on in the local computational graph constructed by Autograd which would remove the memory benefits of sequential aggregation. To avoid that, we disable autograd when doing the distributed message passing step, i.e, the message passing calls (``apply_edges`` and ``update_all``) disable Autograd . We construct this computational graph piece by piece during the backward pass, where each piece involves only the data of two partitions as shown below:

![image](https://user-images.githubusercontent.com/16276103/209869018-36eb38b4-a14a-47ee-940e-5193a67e18e1.png)

As machine 1 receives errors for its local nodes, it follows the following steps:
1. Backpropagate the errors in the local partition.
2. Construct the message passing computational graph involving only machine 1 and machine 2, backpropagtes the errors to machine 2, then delete the computational graph.
3. Construct the message passing computational graph involving only machine 1 and machine 3, backpropagtes the errors to machine 3, then delete the computational graph. 

Machines 2 and 3 follow similar steps. constructing the computational graph piece may involve re-fetching the remote features and running a partial message passing round. For many GNN layer types however, such as GraphSage, GCNII, etc.. we can backpropagate the gradients to remote machines without knowing the remote input features. That is because backpropagating errors through a sum or mean operator does not require that we know the operator's inputs. Our method detects such cases and backpropagates the errors directly without remote re-fetching.

In summary, we obtain linear memory scaling in both the forward and the backward pass by NOT constructing the prohibitively expensive message passing computational graph in the forward pass, but instead constructing it piece by piece during the backward pass. The computational graph for all operations other than the message passing operations is constructed normally during the forward pass.

## API and usage scenarios

We will re-use the same graph partitioning data generated by DistDGL's metis-based partitioning (``dgl.distributed.partition_graph``) to make it easy to switch between sampling-based and full-graph training:

![image](https://user-images.githubusercontent.com/16276103/209869037-fd9584f3-c9fa-4996-a2f8-2082aedffeba.png)

We propose to add a sub-package to DGL: ``sar`` that can be used to run distributed full-graph training and inference as follows:
 
```python

#sar uses torch.distributed for all communication. Backend
#has to support all_to_all collective. 'nccl' backend for GPU training and 'ccl'
#backend for CPU training are recommended
torch.distributed.init_process_group('ccl',rank = rank,world_size = world_size)

#Load the partition data created by dgl.distributed.partition_graph
partition_data = dgl.sar.load_partition_data('path to json file',rank)

full_dist_graph = dgl.sar.construct_graph(partition_data) 

model = make_gnn_model()
model = torch.nn.parallel.DistributedDataParallel(model)

optimizer = torch.optim.Adam(model.parameters(),lr)

model.train()
for train_iter in range(n_train_iters):
    y = model(full_dist_graph,partition_data.node_features['feats'])
    loss = calculate_loss(y,partition_data.node_features['labels'],
			partition_data.node_features['train_mask'])
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    
model.eval()
with torch.no_grad():
     y = model(full_dist_graph,partition_data.node_features['feats'])
     valid_loss = calculate_loss(y,partition_data.node_features['labels'],
			partition_data.node_features['valid_mask'])
```

``full_dist_graph`` exposes a subset of the API of ``dgl.DGLHeteroGraph`` that is relevant for GNN training. In particular it implements a distributed memory-efficient version of ``update_all`` and ``apply_edges``. It can thus be passed directly to a standard DGL GNN.

### distdgl and sar

This diagram highlights the main differences between ``sar`` and ``distdgl``
![image](https://user-images.githubusercontent.com/16276103/209869066-e3ee4084-8e05-44be-b1b7-f028e3e3461b.png)

``sar`` needs to run message passing on the distributed graph so it requires the graph to be compacted, i.e, node indices start from 0 and have no gaps. To enable sequential aggregation, we also need the graph partition at each machine to be split into multiple parts where each part includes the edges from only one remote partition. Finally, full-graph training and inference requires several synchronization points in the forward and backward passes. Our experiments indicate that synchronous all_to_all collectives are faster than point-to-point communication through Tensorpipe.

One usage scenario that we want to make as painless as possible is distributed sampling-based training using ``distdgl``followed by distributed full-graph inference using ``sar``. The simplest way to do that is:

```python

import dgl
import torch as torch

dgl.distributed.initialize('ip_config.txt')
torch.distributed.init_process_group(backend='ccl',rank = rank,world_size = world_size)
g = dgl.distributed.DistGraph('graph_name', 'part_config.json')

# .. rest of distdgl training pipeline ...

##Distributed inference using full graph
partition_data = dgl.sar.load_partition_data('part_config.json',rank)
full_dist_graph = dgl.sar.construct_graph(partition_data) 
model.eval()
with torch.no_grad():
     y = model(full_dist_graph,partition_data.node_features['feats'])
     valid_loss = calculate_loss(y,partition_data.node_features['labels'],
			partition_data.node_features['valid_mask'])

```

Alternatively, we can support direct conversion from ``DistGraph`` to ``GraphShardManager`` to avoid loading the partition data from disk twice.

**Comment by jermainewang** [comment]
Hi @hesham-mostafa , please forgive the delayed reply. We are busy with the v1.0 release and I also spent some time reading your RFC and paper. This is a great write-up. Really appreciate the effort!

At a high-level, I like the proposal to enhance DGL with the capability of distributed full graph computation. The use case of sampling-based mini-batch training + full graph inference is especially interesting. The new capability can also enable large-scale graph diffusion based models in DGL. Examples are the [predict-then-propagate model](https://openreview.net/pdf?id=H1gL-2A9Ym) and the [correct-and-smooth model](https://arxiv.org/abs/2010.13993). The core operation of these models is similar -- propagate a node's embedding to its neighbors and conduct aggregation (and thus can be implemented by `g.update_all`). The difference is that some of the models need the gradients to be propagated backwardly while others do not. It looks like SAR has already supported backward propagation, so it should be able to support both kinds with no problem.

On the technical side, I want to push the effort further -- is it possible to directly support the feature in `DistGraph` instead of creating a new `GraphShardManager`? This will save the maintenance cost moving forward. This may require some refactor/redesign of the current `DistGraph` (which we have been investigating the possibility already). Here are some related questions:

* How does SAR represent the graph? Could you briefly describe the data structure? And why is it more optimized for message passing?
* Does SAR reuse the DistTensor abstraction for storing the input and output of distributed message passing?
* You mentioned that SAR utilize the collective communication primitives from `torch.distributed`. Could you provide more details?

**Comment by jermainewang** [comment]
cc @zheng-da @classicsong

**Comment by hesham-mostafa** [comment]
Hi @jermainewang . Thanks for the detailed response. I agree that the full-graph training/inference capabilities can be used for pre-processing in models like [SIGN](https://arxiv.org/abs/2004.11198) and [SAGN](https://arxiv.org/abs/2104.09376), or for post-processing in techniques such as [correct-and-smooth](https://arxiv.org/abs/2010.13993). If these operations are encapsulated in a ``torch.no_grad()`` or if no input tensors require gradients, then no distributed computational trace will be maintained.
  
  We opted not to incorporate the proposed functionality in DistGraph as there is not much functionality that we can reuse from DistGraph. The graph representation in DistGraph is different from what we need (see point 1 below). If we were to incorporate the full-graph forward/backward functionality in DistGraph, DistGraph would then need to maintain 2 different graph representations that would be used for 2 different purposes: sampling and full-graph message passing. We thought it is cleaner to define a separate class : ``GraphShardManager`` which could be easily derived from DistGraph or initialized directly from the ``GraphPartitionBook``. This modularity would make it easier to optimize/extend distributed sampling-based training (in DistGraph) and full-graph training/inference (in GraphShardManager).
  
  That being said, if you think usability will suffer, then we can think of something to add the full-graph forward/backward functionality in DistGraph while keeping it as separate as possible from the distributed sampling parts.
  
  1)SAR represents the graph in worker ``i`` as a list of ``DGLBlock`` objects where the ``jth`` ``DGLBlock`` in the list represents all the edges coming from vertices in worker/partition ``j`` to vertices in worker/partition ``i``. The figure below right illustrates the 2 ``DGLBlock``s when you have two workers/partitions. 
  
  ![image](https://user-images.githubusercontent.com/16276103/211457538-2718792b-2e9a-43f7-b680-a2bf1fd7c12d.png)
  
  It is optimized for message passing as the node indices in a ``DGLBlock`` have been renumbered to be contiguous and start at 0. If you want to do message passing on the graph representation on the left for example, then you need a feature tensor whose size is the maximum node index which is wasteful as there are only 4 nodes. You have a similar situation in sampling-based training where you call ``dgl.sampling.sample_neighbors`` which yields a non-compacted graph which you then have to pass to ``dgl.to_block`` to obtain a ``DGLBlock`` on which you can efficiently do message passing. Back to SAR, its pseudo-code for implementing GraphSage is :
  
  ```python

#gsm : GraphShardManager object
#gsm.shards[j] is the DGLBlock representing all incoming edges from worker/partition j to the 
#local partition. We have N workers/partitions
gsm.shards = [block_1,...,block_N]
#gsm.num_local_nodes is the number of nodes in this worker's partition
output = torch.zeros(gsm.num_local_nodes,feature_dim)
for src_worker_idx,block in enumerate(gsm.shards):
       #We fetch the neighboring src nodes from worker src_worker_idx
       #using the global node ids in block.srcdata[dgl.NID]
       remote_features = fetch_features(src_worker_idx,block.srcdata[dgl.NID])
       with block.local_scope():
              block.srcdata['h'] = remote_features
              block.update_all(fn.copy_u('h','temp'),fn.sum('temp','result'))
              #All nodes indices in  a partition are contiguous, To map from global indices
              #to indices relative to the start of the local partition, we need to subtract the 
              #global_offset which is the global index of the first node in the local partition
              output[block.dstdata['dgl.NID'] - gsm.global_offset] += block.dstdata['result']
output  = output / gsm.in_degrees()
  
```

2)SAR currently does not use ``DistTensor`` so there is no way for one worker to access the node/edge features in another worker. That is something that definitely needs to change. Modifying SAR to use ``DistTensor`` looks like the best solution.

3)We experimented with point-to-point communication through tensorpipe. However, it was considerably slower than using collective communication through  [``all_to_all``](https://pytorch.org/docs/stable/distributed.html#torch.distributed.all_to_all) from ``torch.distributed``. ``all_to_all`` is supported  by ``NCCL`` (Nvidia's collective communication library) which is part of ``PyTorch`` as well as ``OneCCL`` (Intel's collective communication library) whose PyTorch bindings can be installed from [here](https://github.com/intel/torch-ccl). The code is re-factored so that it is easy to slot in any communication backend that implements the functionality of ``all_to_all``

**Comment by jermainewang** [comment]
From offline discussion:
* In general this is a valuable addition to the DGL ecosystem.
* The design of SAR is currently quite independent of DGL/DistDGL. Users can download and install SAR easily and the corresponding APIs are also well-documented.
* We thus decide to keep the current status of having SAR as a standalone package. To bring the two communities closer, SAR team will create a tutorial about how to use SAR with DGL for scenarios such as full graph training or predict-and-propagate models. DGL team will host the tutorial on its doc site (docs.dgl.ai).

cc @frozenbugs @hesham-mostafa
